dlib/type_traits.h

```diff
@@ -210,8 +210,102 @@ namespace dlib
     template<class T>
     using is_complete_type = details::is_complete_type<T>;
 
- // ----------------------------------------------------------------------------------------
+// ----------------------------------------------------------------------------------------
  
+    template<typename... T>
+    struct types_ {};
+    /*!
+        WHAT THIS OBJECT REPRESENTS
+            This is a type list. You can use this for general-purpose meta-programming
+            and it's used to pass types to the switch_() function.
+    !*/
+
+// ----------------------------------------------------------------------------------------
+
+    namespace details
+    {
+#if defined(__has_builtin)
+#if __has_builtin(__type_pack_element)
+    #define HAS_TYPE_PACK_ELEMENT 1
+#endif
+#endif
+
+#if HAS_TYPE_PACK_ELEMENT
+        template<std::size_t I, class... Ts>
+        struct nth_type { using type = __type_pack_element<I,Ts...>; };
+#else
+        template<std::size_t I, class... Ts>
+        struct nth_type;
+
+        template<std::size_t I, class T0, class... Ts>
+        struct nth_type<I, T0, Ts...> { using type = typename nth_type<I-1, Ts...>::type; };
+
+        template<class T0, class... Ts>
+        struct nth_type<0, T0, Ts...> { using type = T0; };
+#endif
```

Comment:
```suggestion
        template<std::size_t I, class... Ts>
        struct nth_type;

        template<std::size_t I, class T0, class... Ts>
        struct nth_type<I, T0, Ts...> { using type = typename nth_type<I-1, Ts...>::type; };

        template<class T0, class... Ts>
        struct nth_type<0, T0, Ts...> { using type = T0; };
```

Why not this?  If `__type_pack_element` was a "when we get to C++-some-new-version" use this and delete the `#else` block then that's one thing, but `__type_pack_element` is a non-standard language extension isn't it?

Reply:
Yes but it's supported by pretty much every version of clang and later versions of GCC. It's an O(1) lookup so it's the fastest possible implementation. It this were used heavily in template meta-programming then it could make a huge difference. It's also appropriately guarded by the `__has_builtin` macro so it's harmless.

Reply:
We should strive to avoid template recursion as it absolutely kills compilation times.

Reply:
If you disagree with using `__type_pack_element` then we can go without. The implementation is a bit naive but at the moment it's ok because it's only used in `type_safe_union`, which is probably only used for like up to 5 types, and now `callable_traits` which again is only for a handful of types. If it were used by the super heavily templated dnn stuff, then it could be a game changer if we used a proper implementation, or `__type_pack_element` which is the most optimal solution.

Reply:
Na, you are right. Making the build faster is worth it 😁